File: src/test/java/no/nav/fo/veilarbregistrering/registrering/tilstand/RegistreringTilstandServiceTest.java
package no.nav.fo.veilarbregistrering.registrering.tilstand;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import java.time.LocalDateTime;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.*;

public class RegistreringTilstandServiceTest {

    private RegistreringTilstandRepository registreringTilstandRepository;

    private RegistreringTilstandService registreringTilstandService;

    @BeforeEach
    public void setup() {
        registreringTilstandRepository = mock(RegistreringTilstandRepository.class);

        registreringTilstandService = new RegistreringTilstandService(registreringTilstandRepository);
    }

    @Test
    public void skal_oppdatere_registreringtilstand_med_status_og_sistendret() {
        LocalDateTime sistEndret = LocalDateTime.now();
        RegistreringTilstand original = RegistreringTilstandTestdataBuilder
                .registreringTilstand()
                .status(Status.MOTTATT)
                .opprettet(sistEndret.minusDays(1))
                .sistEndret(sistEndret)
                .build();
        when(registreringTilstandRepository.hentRegistreringTilstand(original.getId())).thenReturn(original);

        registreringTilstandService.oppdaterRegistreringTilstand(OppdaterRegistreringTilstandCommand.of(original.getId(), Status.OVERFORT_ARENA));

        ArgumentCaptor<RegistreringTilstand> argumentCaptor = ArgumentCaptor.forClass(RegistreringTilstand.class);
        verify(registreringTilstandRepository).oppdater(argumentCaptor.capture());
        RegistreringTilstand capturedArgument = argumentCaptor.getValue();

        assertThat(capturedArgument.getId()).isEqualTo(original.getId());
        assertThat(capturedArgument.getBrukerRegistreringId()).isEqualTo(original.getBrukerRegistreringId());
        assertThat(capturedArgument.getOpprettet()).isEqualTo(original.getOpprettet());
        assertThat(capturedArgument.getStatus()).isEqualTo(Status.OVERFORT_ARENA);
    }
}

File: src/test/kotlin/no/nav/fo/veilarbregistrering/registrering/formidling/RegistreringTilstandServiceTest.kt
package no.nav.fo.veilarbregistrering.registrering.formidling

import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import no.nav.fo.veilarbregistrering.registrering.formidling.RegistreringTilstandTestdataBuilder.registreringTilstand
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

class RegistreringTilstandServiceTest {

    private lateinit var registreringTilstandRepository: RegistreringTilstandRepository
    private lateinit var registreringTilstandService: RegistreringTilstandService

    @BeforeEach
    fun setup() {
        registreringTilstandRepository = mockk()
        registreringTilstandService = RegistreringTilstandService(registreringTilstandRepository)
    }

    @Test
    fun `skal oppdatere registreringtilstand med status og sistendret`() {
        val sistEndret = LocalDateTime.now()
        val original = registreringTilstand()
            .status(Status.MOTTATT)
            .opprettet(sistEndret.minusDays(1))
            .sistEndret(sistEndret)
            .build()

        every { registreringTilstandRepository.hentRegistreringTilstand(original.id) } returns original
        val registreringTilstandSlot = slot<RegistreringTilstand>()
        every { registreringTilstandRepository.oppdater(capture(registreringTilstandSlot)) } returns original

        registreringTilstandService.oppdaterRegistreringTilstand(
            OppdaterRegistreringTilstandCommand.of(
                original.id,
                Status.OVERFORT_ARENA
            )
        )

        val capturedArgument = registreringTilstandSlot.captured
        assertThat(capturedArgument.id).isEqualTo(original.id)
        assertThat(capturedArgument.brukerRegistreringId).isEqualTo(original.brukerRegistreringId)
        assertThat(capturedArgument.opprettet).isEqualTo(original.opprettet)
        assertThat(capturedArgument.status).isEqualTo(Status.OVERFORT_ARENA)
    }
}
